### PR TITLE
Insert and render metrics in real-time

### PR DESCRIPTION
**Pull request by thorrester**
### Overview

This PR adds the ability to log `metrics` during an active run. This is important for long-running jobs where a DS may wish to view ongoing metrics in the UI. 

Change overview:

- Adds new `metrics` table in DB for uploading and downloading of metrics
- Adds `log_graph` method for logging x, y graphs in real-time with UI visualization
- Changes `load_preprocessor` default to `false`
- Updates to UI javascript for graph visualization

Future TODO:

- revisit methods involving metrics (uploading, downloading, champion/challenger, ui logic) to see if there are ways to simplify. There may not be, but it's worth a look in another PR

### CODE REVIEW

File: opsml/cards/run.py [hunk 1]
```diff
@@ -359,6 +544,12 @@ def uri(self) -> Path:
             end_path,
         )
 
+    @cached_property
+    def _registry(self) -> RunCardRegistry:
+        from opsml.registry.backend import _set_registry
+
+        return cast(RunCardRegistry, _set_registry(RegistryType.RUN))
+
```
Comment (thorrester) [inline opsml/cards/run.py:552]: this slightly breaks our pattern of keeping cards and registries separate; however it's needed to record metrics during an active `run`

File: opsml/storage/api.py [hunk 2]
```diff
@@ -31,6 +31,8 @@ class ApiRoutes:
     MODEL_METADATA = "models/metadata"
     MODEL_METRICS = "models/metrics"
     PROJECT_ID = "projects/id"
+    UPLOAD_METRICS = "metrics/upload"
+    DOWNLOAD_METRICS = "metrics/download"
```
Comment (thorrester) [inline opsml/storage/api.py:35]: using `upload` and `download` for consistency with `files` pathing

**Comment by thorrester** [comment]
Users will now be able to log graphs based on x,y data at runtime

graph

File: examples/huggingface/hf_example.py [hunk 3]
```diff
@@ -159,7 +159,7 @@ def _test_onnx_model(self):
         modelcard: ModelCard = self.registries.model.load_card(name=self.info.name)
 
         # load onnx model
-        modelcard.load_onnx_model()
+        modelcard.load_onnx_model(load_preprocessor=True)
```
Comment (damonallison) [inline examples/huggingface/hf_example.py:162]: Good catch.

File: opsml/app/routes/metrics.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+@router.post("/metrics", response_model=MetricUploadResponse, name="metric_post")
+def insert_metric(request: Request, payload: MetricsModel) -> MetricUploadResponse:
+    """Inserts metrics into metric table
+
+    Args:
+        request:
+            FastAPI request object
+        payload:
+            MetricsUploadRequest
```
Comment (damonallison) [inline opsml/app/routes/metrics.py:28]: MetricsModel - or we remove these comments unless being flagged by a linter as they aren't adding any value.

File: opsml/app/routes/metrics.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    """
+
+    run_reg: ServerRunCardRegistry = request.app.state.registries.run._registry
+
+    try:
+        metrics = run_reg.get_metric(run_uid, name)
+        return MetricsModel(metric=metrics)
+
+    except Exception as error:
+        logger.error(f"Failed to insert metrics: {error}")
```
Comment (damonallison) [inline opsml/app/routes/metrics.py:70]: insert -> get?

File: opsml/app/routes/pydantic_models.py [hunk 6]
```diff
@@ -206,15 +206,20 @@ class DeleteFileRequest(BaseModel):
     read_path: str
 
 
-class MetricRequest(BaseModel):
-    name: Optional[str] = None
-    repository: Optional[str] = None
-    version: Optional[str] = None
-    uid: Optional[str] = None
+class MetricUploadResponse(BaseModel):
```
Comment (damonallison) [inline opsml/app/routes/pydantic_models.py:209]: When would this be false? Maybe we just return a 200?

File: opsml/app/routes/pydantic_models.py [hunk 7]
```diff
@@ -206,15 +206,20 @@ class DeleteFileRequest(BaseModel):
...
+    uploaded: bool
+
+
+class MetricModel(BaseModel):
```
Comment (damonallison) [inline opsml/app/routes/pydantic_models.py:213]: Surprised we didn't have this until now :)

File: opsml/projects/active_run.py [hunk 8]
```diff
@@ -283,6 +286,59 @@ def log_parameters(self, parameters: Dict[str, Union[float, int, str]]) -> None:
         self._verify_active()
         self.runcard.log_parameters(parameters=parameters)
 
+    def log_graph(
```
Comment (damonallison) [inline opsml/projects/active_run.py:289]: Is there any 3rd party library we could use as the basis for our graphing functionality? This is a potentially big space we're opening up here - for example we may want to plot multiple values on the same graph, different graph types, etc.

Let's chat about an overall plan offline.

**Review by damonallison — APPROVED** [review]

File: opsml/app/routes/metrics.py (hunk 4 above)
Comment (thorrester) [reply opsml/app/routes/metrics.py:28]: Good catch. I'll fix the docstrings

File: opsml/app/routes/metrics.py (hunk 5 above)
Comment (thorrester) [reply opsml/app/routes/metrics.py:70]: ahhh!

File: opsml/cards/run.py [hunk 9]
```diff
@@ -18,16 +26,116 @@
...
+    Args:
+        array:
+            List of floats or ints
+
+    Returns:
+        Decimated array
+    """
+    length = len(array)
+    if len(array) > 200_000:
+        step = round(length / 200)
```
Comment (thorrester) [inline opsml/cards/run.py:82]: Need to change this to `200_000`

File: opsml/app/routes/pydantic_models.py (hunk 6 above)
Comment (thorrester) [reply opsml/app/routes/pydantic_models.py:209]: Good call. We can return a 200

**Pull request merged**